### PR TITLE
Introducing Feature Rapid Rebalance

### PR DESCRIPTION
**Pull request by ziggie1984**
This PR introduces the possibility to enable rapid rebalance, which means that as soon as a successful route is found, it will use this route and those 2 channels (from, to) to rebalance until either the route is depleted or the channel do not meet the starting criteria anymore.

Whats are your thoughts about this feature, should we include it in the tool, or should we lets people build wrapper to achieve exactly that?

Thanks for looking into it

### CODE REVIEW

**Comment by ziggie1984** [comment]
It also applies to probing, as soon as a probing amount fits, it will use this new amount for rapid rebalance if the new feature is selected

**Comment by rkfg** [comment]
Thanks for the feature, sounds really good! I haven't looked much into the code but noticed there's a lot of duplicating logic between the new function and the existing ones. Maybe it's possible to unify this so we wouldn't have to change the channel selection logic in two places if there's any additions in the future? For example, regenerate the channel pairs from scratch, check if the pair we're interested in is still in the map. It shouldn't take much time (m*n map inserts where m and n are the number of incoming and outgoing channels) but the code will be easier to maintain.

**Comment by ziggie1984** [comment]
totally agree, not satisfied with the duplicated code as well, will unify just wanted to get the feedback whether this feature will make it

**Comment by rkfg** [comment]
Oh, absolutely! It's useful and we've discussed it with bitcoinite. Also, there's a conflict that needs to be resolved, I suppose related to the CLI parameter rename that I did because it didn't match the config keys.

**Comment by ziggie1984** [comment]
ok will prepare everything today, so you can have a look at the weekend !

**Comment by ziggie1984** [comment]
Nice one, thanks for the efficiency tip, will use your recommendations

File: main.go [hunk 1]
```diff
@@ -191,6 +215,98 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
...
+
+		r.channels = r.channels[:0]
+		r.channels = append(r.channels, toChan.Channels...)
+		r.channels = append(r.channels, fromChan.Channels...)
+
+		for k, _ := range r.failureCache {
+			delete(r.failureCache, k)
+		}
+
+		err = r.getChannelCandidates(params.FromPerc, params.ToPerc, params.Amount)
```
Comment (rkfg) [inline main.go:269]: Shouldn't the last parameter be `amt`?

File: main.go [hunk 2]
```diff
@@ -191,6 +215,98 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
 	return nil, true
 }
 
+func tryRapidRebalance(ctx context.Context, r *regolancer, from, to uint64, route *lnrpc.Route, amt int64) (successfullAtempts int, err error) {
+
+	rapidAttempt := 1
```
Comment (rkfg) [inline main.go:220]: A nitpick but I'd set it to 0 and only add 1 where it's needed. In most places below you subtract one. Not a strong opinion, I see that it makes sense to start with 1.

File: main.go [hunk 3]
```diff
@@ -191,6 +215,98 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
...
+
+		cTo, err := r.getChanInfo(ctx, to)
+
+		if err != nil {
+			return rapidAttempt - 1, err
+		}
+		cFrom, err := r.getChanInfo(ctx, from)
+
+		if err != nil {
+			logErrorF("rapid balance error: %s", err)
```
Comment (rkfg) [inline main.go:234]: No idea what errors can be here but making them slightly more informative wouldn't hurt! "rapid balance error" in every "iferrnil" isn't informative enough.

**Review by rkfg — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (ziggie1984) [reply main.go]: good catch

File: main.go (hunk 2 above)
Comment (ziggie1984) [reply main.go]: is more elegant I guess

File: main.go (hunk 3 above)
Comment (ziggie1984) [reply main.go]: added some more verbose information

File: main.go [hunk 4]
```diff
@@ -164,10 +167,21 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
 	routeCtxCancel()
 	for _, route := range routes {
 		log.Printf("Attempt %s, amount: %s (max fee: %s sat | %s ppm )",
-			hiWhiteColorF("#%d", *attempt), hiWhiteColor(amt), formatFee(fee), formatFeePPM(amt*1000, fee))
+			hiWhiteColorF("#%d", *attempt), hiWhiteColor(amt), formatFee(fee), formatFeePPM(amt*1000-fee, fee))
```
Comment (rkfg) [inline main.go:170]: This change is interesting. I'm personally not sure whether we should display the total amount with fees (that's being spent) or without (that we receive in target channel). Either way, if we change it here the success message should be changed too. I suggest we make it a separate PR, it doesn't affect the logic, only the presentation.

**Review by rkfg — CHANGES_REQUESTED** [review]

File: main.go [hunk 5]
```diff
@@ -191,6 +215,99 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
 	return nil, true
 }
 
+func tryRapidRebalance(ctx context.Context, r *regolancer, from, to uint64, route *lnrpc.Route, amt int64) (successfullAtempts int, err error) {
+
+	rapidAttempt := 0
+
+	for {
+
+		log.Printf("Rapid rebalance attempt %s", hiWhiteColor(rapidAttempt))
```
Comment (rkfg) [inline main.go:224]: Here we need hiWhiteColor(rapidAttempt+1) if we count from zero

**Review by rkfg — CHANGES_REQUESTED** [review]

File: main.go (hunk 4 above)
Comment (ziggie1984) [reply main.go:170]: ohh sry, was making some changes on dev branch, that got through, will revert.

Comment (ziggie1984) [reply main.go]: Yeah, I think we should unify that in another PR, bc I am calculating the success message taking the fees into account, so its even not correct what I did, I should have added the fees to be consistent with the success message

File: main.go (hunk 5 above)
Comment (ziggie1984) [reply main.go:224]: ok will squash some commits now, getting too much lol

Comment (rkfg) [reply main.go:224]: Indeed, just one is enough!

Comment (ziggie1984) [reply main.go]: ok will do

**Comment by rkfg** [comment]
The code looks great! I'll test it soon. Also, I think it's a good tradition to send a tip for good PRs, post your Lightning address. You've added many nice features already.

**Comment by Bitcoinite** [comment]
Thanks for your work on this PR. I really appreciate how quickly you have implemented this idea.

**Comment by ziggie1984** [comment]
Thanks for your help :)

ziggie1984@lnme.ziggie-dev.org

**Comment by ziggie1984** [comment]
added option description to README and removed false comment.

Thanks for donation :)

**Comment by rkfg** [comment]
Not quite sure what's going on here. The command was `regolancer -l --allow-rapid-rebalancing --allow-unbalance-from --rel-amount-to=0.3` (some parameters are also set in the config such as 5 probe steps and 50000 min amount). The result:


The target channel went up to 62% even though the limit should be 50% (I only allowed to unbalance the source channel). I'm also unsure how we should deal with `--pto` for example. It affects what channels are chosen but should it be respected during rapid fire? What about `--rel-amount-to`? It's used for amount calculation but if rapid fire is used it'd be mostly useless as either the goal is reached instantly (and then no rapid fire should be possible) or probing fails as I described above and we use min amount until we reach the relative amount goal (rare occurrence, if min amount probing succeeded in most cases a higher amount is discovered which makes rapid fire impossible). What do you think?

**Comment by ziggie1984** [comment]
hmm interesting why your local got over 62%, lets me check

**Comment by ziggie1984** [comment]
hmm I guess we will still need `pto`, because how would we otherwise check, that the rapid fire did reach the limit and should stop?

regarding `rel-amount-to` I guess this makes no sense to rapid rebalance in that case and the program should stop instead of trying an amount which will not succeed anyway

**Comment by rkfg** [comment]
Agree, relative amounts should disable rapid fire. It can be added to the preflight checks. There are many of them already, might make sense to extract them to a separate func maybe? I wonder why `pto` didn't work though.

**Comment by ziggie1984** [comment]
Correct,

but still strange why the rebalance did not stop regarding the `--rel-amount-to=0.3 ` though, normally maxTo amount should go to zero and then should exit with `no_routes` ...

**Comment by ziggie1984** [comment]
> There are many of them already, might make sense to extract them to a separate func maybe

Yeah good idea, you mean alle the params check in main(), to export in a function preflightCheck(), checking for rel amount and then disabling rapid-rebalance, including moving all the existing ones ?

**Comment by rkfg** [comment]
Yeah, all that long list of setting defaults and checking if options don't contradict each other. The piece between parsing the config and establishing lnd connection. Nothing feels as good as doing a safe and quick refactoring so that things are less messy.

**Comment by ziggie1984** [comment]
ok I might found the problem, I guess when wiping all channels and adding them with `getchannelcandidate`, but I guess here things start to become messy, I think we need to really add the `to` and `from` channel expliciltly to prevent `to` and `from` being switched after some rebalances

**Comment by ziggie1984** [comment]
I think I will wipe `fromChannelId` and `toChannelId` then add the from and to channel to them and then call getchannelcandidate, that would prevent just messy behaviour, wdyt ?

**Comment by ziggie1984** [comment]
ok works now, had to also wipe fromChannel and toChannel lol

File: main.go [hunk 6]
```diff
@@ -191,6 +214,115 @@ func tryRebalance(ctx context.Context, r *regolancer, attempt *int) (err error,
...
+		r.fromChannelId = makeChanSet([]uint64{from})
+		for k, _ := range r.toChannelId {
+			delete(r.toChannelId, k)
+		}
+
+		r.toChannelId = makeChanSet([]uint64{to})
+
+		r.channels = r.channels[:0]
+		r.fromChannels = r.fromChannels[:0]
+		r.fromChannels = r.fromChannels[:0]
```
Comment (rkfg) [inline main.go:274]: Looks like it shoud be `r.toChannels` here

**Review by rkfg — CHANGES_REQUESTED** [review]

**Comment by ziggie1984** [comment]
should we even add more to the preflight checks ?

**Comment by rkfg** [comment]
I think it's better to error out if we specify rapid rebalance and relative amounts instead of turning it off and continuing. Explicit errors are better so that the user knows what to expect and how to deal with it. Less surprise factors.

**Comment by Bitcoinite** [comment]
> I think it's better to error out if we specify rapid rebalance and relative amounts instead of turning it off and continuing. Explicit errors are better so that the user knows what to expect and how to deal with it. Less surprise factors.

I'm running regolancer as a bash script in the background, so I don't see any of the outputs and won't notice errors like this

**Comment by rkfg** [comment]
> I'm running regolancer as a bash script in the background, so I don't see any of the outputs and won't notice errors like this

That's fine but it still has a user-friendly output. It's better to report to the user if they do something unexpected or wrong so they don't have false expectation.

**Comment by ziggie1984** [comment]
> > I'm running regolancer as a bash script in the background, so I don't see any of the outputs and won't notice errors like this
> 
> That's fine but it still has a user-friendly output. It's better to report to the user if they do something unexpected or wrong so they don't have false expectation.

Could he maybe just errors of regolancer in his bashscript by doing if [$?], if this occurs he should investigate his cronjobs

**Comment by rkfg** [comment]
> > > I'm running regolancer as a bash script in the background, so I don't see any of the outputs and won't notice errors like this
> > 
> > 
> > That's fine but it still has a user-friendly output. It's better to report to the user if they do something unexpected or wrong so they don't have false expectation.
> 
> Could he maybe just check for errors of regolancer in his bashscript by doing if [$?], if this occurs he should investigate his cronjobs

Yep, shouldn't be hard to detect and report them. Or just add `-e` to the first shebang line so the whole script terminates if any command returns a non-zero code.

**Comment by rkfg** [comment]
Looks like something broke in this PR. I don't have any connection params set (address, macaroon, tls etc.) and it works in master but exits on this branch with this message:
```
2022/10/29 00:47:05 read /home/pi/.lnd/data/chain/bitcoin: is a directory
```
Quite weird.

**Comment by rkfg** [comment]
Ah, got it. You're passing params to preflightChecks by value so all changes don't apply to the actual params var. Pass it by pointer and it works. Did you test it? 👀

**Comment by ziggie1984** [comment]
Oh shit, yes I am running the final version though everything is in my config so I did not see the error!

**Comment by rkfg** [comment]
Another interesting occurrence

I think it happened because the balance went above `--pto=30` (it's at 32% currently), the channel capacity is 4kk so 100k is 2.5%. It was selected as a target because it was at 29.7% (I calculated the previous balance manually, there were no forward involving this channel so it's accurate), then it went above the target. Which resulted in no target channels. This error is confusing but I think it's not the patch fault.

Will do a few more tests and if everything's fine it's good to go!

**Review by rkfg — APPROVED** [review]

**Comment by ziggie1984** [comment]
> Another interesting occurrence  I think it happened because the balance went above `--pto=30` (it's at 32% currently), the channel capacity is 4kk so 100k is 2.5%. It was selected as a target because it was at 29.7% (I calculated the previous balance manually, there were no forwards involving this channel so it's accurate), then it went above the target. Which resulted in no target channels. This error is confusing but I think it's not the patch fault.
> 
> Will do a few more tests and if everything's fine it's good to go!

Agree found this error a bit misleading, would be cooler to just fail saying `channel already in balance or channel criteria not met anymore ...`, but bc we are using the general methods for channel selection as you mentioned correctly we could tackle this one in a seperate PR maybe ?

**Comment by rkfg** [comment]
Yep, let's do that. I'm still running tests but even with 10k it's hard to find a route that would let me send more than once...

**Comment by ziggie1984** [comment]
I tested the rapid rebalance by doing unprofitable rebalances, though yesterday I was able to send 40 rapid rebalances through the WOS node

Screenshot 2022-10-29 at 21 38 28

**Comment by rkfg** [comment]
All good! Thanks for working on it!

**Comment by ziggie1984** [comment]
thanks for your detailed support, appreciate it !
